docs/man/man3/TTT_Log.3
.TH "TTT::Log" 3 "Thu Mar 9 2023" "Version v1.0.1" "TTT" \" -*- nroff -*-
.ad l
.nh
.SH NAME
TTT::Log \- Logger class not really meant to be used outside of the library\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include 'TTT\&.hpp'\fP
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static void \fBlog\fP (\fBmodule\fP \fBmodule\fP, \fBlevel\fP \fBlevel\fP, std::string message)"
.br
.RI "Function to log the debug output to the terminal and/or the 'TTT\&.log' file\&. "
.in -1c
.SS "Static Public Attributes"

.in +1c
.ti -1c
.RI "static const \fBmodule\fP \fBTTT\fP = 'TTT'"
.br
.RI "Module used in the most critical of logs\&. "
.ti -1c
.RI "static const \fBmodule\fP \fBWINDOW\fP = 'WINDOW'"
.br
.RI "Module used in logging everything that has to do with the window\&. "
.ti -1c
.RI "static const \fBmodule\fP \fBUTILS\fP = 'UTILS'"
.br
.RI "Module used in logging everything that has to do with the utilities class\&. "
.ti -1c
.RI "static const \fBmodule\fP \fBLOG\fP = 'LOG'"
.br
.RI "Module used in logging everything that has to do with the logger class\&. "
.ti -1c
.RI "static const \fBmodule\fP \fBSFML\fP = 'SFML'"
.br
.RI "Module used in logging everything that has to do with the SFML library\&. "
.ti -1c
.RI "static const \fBlevel\fP \fBSUCCESS\fP = 'SUCCESS'"
.br
.RI "Success level indicates that an operation was successful\&. "
.ti -1c
.RI "static const \fBlevel\fP \fBINFO\fP = 'INFO'"
.br
.RI "Info level displays basic information\&. "
.ti -1c
.RI "static const \fBlevel\fP \fBWARNING\fP = 'WARNING'"
.br
.RI "Warning level indicates that there's something wrong with the program but doesn't require immediate attention\&. "
.ti -1c
.RI "static const \fBlevel\fP \fBERROR\fP = 'ERROR'"
.br
.RI "Error level indicates that there's something wrong with the program requires attention\&. "
.ti -1c
.RI "static const \fBlevel\fP \fBCRITICAL\fP = 'CRITICAL'"
.br
.RI "Critical level indicates that there's something wrong with the program and requires immediate attention\&. "
.ti -1c
.RI "static const \fBcolor\fP \fBRESET\fP = '\\033[0m'"
.br
.RI "Reset color resets all of the colors to the terminal default\&. "
.ti -1c
.RI "static const \fBcolor\fP \fBRED\fP = '\\033[31m'"
.br
.RI "Red color paints output red\&. Used in critical errors\&. "
.ti -1c
.RI "static const \fBcolor\fP \fBGREEN\fP = '\\033[32m'"
.br
.RI "Green color paints output green\&. Used when something is successful\&. "
.ti -1c
.RI "static const \fBcolor\fP \fBWHITE\fP = '\\033[37m'"
.br
.RI "White color paints output white\&. Used when displaying basic information\&. "
.ti -1c
.RI "static const \fBcolor\fP \fBL_RED\fP = '\\033[91m'"
.br
.RI "Light red color paints output light red\&. Used in errors\&. "
.ti -1c
.RI "static const \fBcolor\fP \fBL_YELLOW\fP = '\\033[93m'"
.br
.RI "Light yellow color paints output light yellow\&. used in warnings\&. "
.in -1c
.SH "Detailed Description"
.PP 
Logger class not really meant to be used outside of the library\&. 
.PP
Definition at line \fB54\fP of file \fBTTT\&.hpp\fP\&.
.SH "Member Function Documentation"
.PP 
.SS "void TTT::Log::log (\fBmodule\fP module, \fBlevel\fP level, std::string message)\fC [static]\fP"

.PP
Function to log the debug output to the terminal and/or the 'TTT\&.log' file\&. 
.PP
\fBParameters\fP
.RS 4
\fImodule\fP module representing the file or library the info/error/success was noted in\&. 
.br
\fIlevel\fP level representing the level of severity that was logged\&. 
.br
\fImessage\fP a string representing the message associated with the log\&. 
.RE
.PP

.PP
Definition at line \fB52\fP of file \fBTTT_Log\&.cpp\fP\&.
.SH "Member Data Documentation"
.PP 
.SS "const \fBlevel\fP TTT::Log::CRITICAL = 'CRITICAL'\fC [static]\fP"

.PP
Critical level indicates that there's something wrong with the program and requires immediate attention\&. 
.PP
Definition at line \fB70\fP of file \fBTTT\&.hpp\fP\&.
.SS "const \fBlevel\fP TTT::Log::ERROR = 'ERROR'\fC [static]\fP"

.PP
Error level indicates that there's something wrong with the program requires attention\&. 
.PP
Definition at line \fB69\fP of file \fBTTT\&.hpp\fP\&.
.SS "const \fBcolor\fP TTT::Log::GREEN = '\\033[32m'\fC [static]\fP"

.PP
Green color paints output green\&. Used when something is successful\&. 
.PP
Definition at line \fB74\fP of file \fBTTT\&.hpp\fP\&.
.SS "const \fBlevel\fP TTT::Log::INFO = 'INFO'\fC [static]\fP"

.PP
Info level displays basic information\&. 
.PP
Definition at line \fB67\fP of file \fBTTT\&.hpp\fP\&.
.SS "const \fBcolor\fP TTT::Log::L_RED = '\\033[91m'\fC [static]\fP"

.PP
Light red color paints output light red\&. Used in errors\&. 
.PP
Definition at line \fB76\fP of file \fBTTT\&.hpp\fP\&.
.SS "const \fBcolor\fP TTT::Log::L_YELLOW = '\\033[93m'\fC [static]\fP"

.PP
Light yellow color paints output light yellow\&. used in warnings\&. 
.PP
Definition at line \fB77\fP of file \fBTTT\&.hpp\fP\&.
.SS "const \fBmodule\fP TTT::Log::LOG = 'LOG'\fC [static]\fP"

.PP
Module used in logging everything that has to do with the logger class\&. 
.PP
Definition at line \fB63\fP of file \fBTTT\&.hpp\fP\&.
.SS "const \fBcolor\fP TTT::Log::RED = '\\033[31m'\fC [static]\fP"

.PP
Red color paints output red\&. Used in critical errors\&. 
.PP
Definition at line \fB73\fP of file \fBTTT\&.hpp\fP\&.
.SS "const \fBcolor\fP TTT::Log::RESET = '\\033[0m'\fC [static]\fP"

.PP
Reset color resets all of the colors to the terminal default\&. 
.PP
Definition at line \fB72\fP of file \fBTTT\&.hpp\fP\&.
.SS "const \fBmodule\fP TTT::Log::SFML = 'SFML'\fC [static]\fP"

.PP
Module used in logging everything that has to do with the SFML library\&. 
.PP
Definition at line \fB64\fP of file \fBTTT\&.hpp\fP\&.
.SS "const \fBlevel\fP TTT::Log::SUCCESS = 'SUCCESS'\fC [static]\fP"

.PP
Success level indicates that an operation was successful\&. 
.PP
Definition at line \fB66\fP of file \fBTTT\&.hpp\fP\&.
.SS "const \fBmodule\fP TTT::Log::TTT = 'TTT'\fC [static]\fP"

.PP
Module used in the most critical of logs\&. 
.PP
Definition at line \fB59\fP of file \fBTTT\&.hpp\fP\&.
.SS "const \fBmodule\fP TTT::Log::UTILS = 'UTILS'\fC [static]\fP"

.PP
Module used in logging everything that has to do with the utilities class\&. 
.PP
Definition at line \fB62\fP of file \fBTTT\&.hpp\fP\&.
.SS "const \fBlevel\fP TTT::Log::WARNING = 'WARNING'\fC [static]\fP"

.PP
Warning level indicates that there's something wrong with the program but doesn't require immediate attention\&. 
.PP
Definition at line \fB68\fP of file \fBTTT\&.hpp\fP\&.
.SS "const \fBcolor\fP TTT::Log::WHITE = '\\033[37m'\fC [static]\fP"

.PP
White color paints output white\&. Used when displaying basic information\&. 
.PP
Definition at line \fB75\fP of file \fBTTT\&.hpp\fP\&.
.SS "const \fBmodule\fP TTT::Log::WINDOW = 'WINDOW'\fC [static]\fP"

.PP
Module used in logging everything that has to do with the window\&. 
.PP
Definition at line \fB61\fP of file \fBTTT\&.hpp\fP\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for TTT from the source code\&.


CHANGELOG.md
[TOC]

# Release Notes
These are the release notes for each new version of TTT.

## v1.0.1 | Mar. 7, 2023
**Added**
- New static variable `defaultEventLoopFontFilename`

## v1.0.0 | Mar. 5, 2023
- Initial release


src/TTT_Log.cpp
/** \file TTT_Log.cpp
 * \brief Defines the "Log" class.
 */

#include "include/TTT/TTT.hpp"

#include <iostream>
#include <fstream>

namespace TTT
{
    const module Log::TTT = "TTT";

    const module Log::WINDOW = "WINDOW";
    const module Log::UTILS  = "UTILS";
    const module Log::LOG    = "LOG";
    const module Log::SFML   = "SFML";

    const level Log::SUCCESS  = "SUCCESS";
    const level Log::INFO     = "INFO";
    const level Log::WARNING  = "WARNING";
    const level Log::ERROR    = "ERROR";
    const level Log::CRITICAL = "CRITICAL";

    const color Log::RESET    = "\033[0m";
    const color Log::RED      = "\033[31m";
    const color Log::GREEN    = "\033[32m";
    const color Log::WHITE    = "\033[37m";
    const color Log::L_RED    = "\033[91m";
    const color Log::L_YELLOW = "\033[93m";

    const std::string currentTime()
    {
        time_t now = time(0);
        struct tm    tstruct;
        char         buf[80];
        tstruct    = *localtime(&now);

        strftime(buf, sizeof(buf), "%X", &tstruct);

        return buf;
    }

    std::fstream logFile;

    /** \brief Function to log the debug output to the terminal and/or the "TTT.log" file.
     *
     * \param module module representing the file or library the info/error/success was noted in.
     * \param level level representing the level of severity that was logged.
     * \param message a string representing the message associated with the log.
     */
    void Log::log(module module, level level, std::string message)
    {
        logFile.open("TTT.log", logFile.app);

        logFile << "TTT - [" << module << " | " << level << "] (" << currentTime() << "): " << message << std::endl;

        if (Window::verbose)
        {
            if (level == Log::SUCCESS)
                std::cout << Log::GREEN    << "TTT - [" << module << " | " << level << "] (" << currentTime() << "): " << message << Log::RESET << std::endl;
            else if (level == Log::INFO)
                std::cout << Log::WHITE    << "TTT - [" << module << " | " << level << "] (" << currentTime() << "): " << message << Log::RESET << std::endl;
            else if (level == Log::WARNING)
                std::cout << Log::L_YELLOW << "TTT - [" << module << " | " << level << "] (" << currentTime() << "): " << message << Log::RESET << std::endl;
            else if (level == Log::ERROR)
                std::cout << Log::L_RED    << "TTT - [" << module << " | " << level << "] (" << currentTime() << "): " << message << Log::RESET << std::endl;
            else if (level == Log::CRITICAL)
                std::cout << Log::RED      << "TTT - [" << module << " | " << level << "] (" << currentTime() << "): " << message << Log::RESET << std::endl;
        }

        logFile.close();
    }
}
